pkg/collector/corechecks/snmp/internal/report/report_bandwidth_usage.go

```diff
@@ -97,18 +98,76 @@ func (ms *MetricSender) sendBandwidthUsageMetric(symbol profiledefinition.Symbol
 	}
 	usageValue := ((octetsFloatValue * 8) / (float64(ifSpeed))) * 100.0
 
-	sample := MetricSample{
-		value:      valuestore.ResultValue{SubmissionType: profiledefinition.ProfileMetricTypeCounter, Value: usageValue},
-		tags:       tags,
-		symbol:     profiledefinition.SymbolConfig{Name: usageName + ".rate"},
-		forcedType: profiledefinition.ProfileMetricTypeCounter,
-		options:    profiledefinition.MetricsConfigOption{},
+	interfaceID := ms.deviceID + ":" + fullIndex + "." + usageName
+	err = ms.calculateRate(interfaceID, ifSpeed, usageValue, usageName, tags)
+	if err != nil {
+		return err
 	}
+	return nil
+}
+
+/*
+calculateRate is responsible for checking the state for previously seen metric sample to generate the rate from.
+If the ifSpeed has changed for the interface, the rate will not be submitted (drop the previous sample)
+*/
+func (ms *MetricSender) calculateRate(interfaceID string, ifSpeed uint64, usageValue float64, usageName string, tags []string) error {
+	ms.interfaceRateMap.mu.RLock()
+	interfaceRate, ok := ms.interfaceRateMap.rates[interfaceID]
+	ms.interfaceRateMap.mu.RUnlock()
```

Comment:
I'm wondering if instead of having a `InterfaceRateMap` object/structure, maybe we can rename it to something like `InterfaceBandwidthState` that encapsulate the calculation of the bandwidth and the states needed (previous values)

`InterfaceBandwidthState`, similar to current InterfaceRateMap, it will have the:
- bandwidth usage value (currently `rates map[string]InterfaceRate`)
- mutex `mu    sync.RWMutex`

But it will have handle the behaviours:
- `CalculateBandwithUsage()` (aka `calculateRate`)

This way, the map of previous values and the mutex will be encapsulated within the `InterfaceBandwidthState` object/structure.

Reply:
so something like renaming the structs like so and then having `InterfaceBandwidthState` handle `calculateBandwidthUsage` instead of having the `MetricSender` do so i think makes sense (if i'm interpreting what you're saying correctly :) )

```
// BandwidthUsage tracks the interface's current ifSpeed and last seen sample to generate rate with
type BandwidthUsage struct {
	ifSpeed        uint64
	previousSample float64
	previousTsNano int64
}

// InterfaceBandwidthState holds state between runs to be able to calculate rate and know if the ifSpeed has changed
type InterfaceBandwidthState struct {
	state map[string]BandwidthUsage
	mu    sync.RWMutex
}
```